projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/lineage_utils.py

```diff
@@ -0,0 +1,55 @@
+# Copyright 2021 VMware, Inc.
+# SPDX-License-Identifier: Apache-2.0
+import logging
+import re
+from typing import Optional
+from typing import Tuple
+
+from impala._thrift_gen.RuntimeProfile.ttypes import TRuntimeProfileFormat
+from impala.hiveserver2 import HiveServer2Cursor
+from vdk.api.lineage.model.sql.model import LineageData
+from vdk.api.lineage.model.sql.model import LineageTable
+
+_log = logging.getLogger(__name__)
+
+
+def get_lineage_data(cursor: HiveServer2Cursor) -> Optional[LineageData]:
+    """
+    Calculates lineage information from the query plan for the last executed query from the provided cursor. Lineage
+    is calculated only if the query has finished successfully and has resulted actual data read or write from/to the
+    underlying storage.
+    """
+    query_profile = cursor.get_profile(profile_format=TRuntimeProfileFormat.STRING)
+    if "Query Status: OK" not in query_profile:
+        return None  # do not capture lineage for failed queries
+    inputs, output = _parse_inputs_outputs(query_profile)
+    if not inputs and not output:
+        return None  # no lineage was present in the query plan, e.g. select 1 query
+    return LineageData(
+        query=cursor._cursor.query_string,  # a bit hacky but otherwise it is not trivial to parse the profile string
+        query_status="OK",
+        query_type=None,  # TODO specify query_type once there is clear spec for it
+        input_tables=[
+            _get_lineage_table_from_table_name(table_name) for table_name in inputs
+        ],
+        output_table=_get_lineage_table_from_table_name(output),
+    )
```

Comment:
does it make sense to add a debug level message that contains the query profile retrieved, so it's troubleshooted easier?

Reply:
I though about it, but for complex queries it is really long like >2k lines. I was thinking to log at least the query id or something if an error occurs, but it does not improve much debug-ability as the query itself is logged and it can be used to retrieve the profile from impala.
What is your opinion?

Reply:
I don't mind